lib/services/reconciler.go

```diff
@@ -147,7 +147,7 @@ func (r *GenericReconciler[K, T]) processRegisteredResource(ctx context.Context,
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	r.logger.InfoContext(ctx, "Resource was removed, deleting", "name", key)
+	r.logger.InfoContext(ctx, "Resource was removed, deleting", "kind", kind, "name", key)
```

Comment:
The changes look great overal. I do  however  have a small suggestion for improvement that could make the implementation a bit cleaner and more concise:

```go
// NewGenericReconciler creates a new GenericReconciler with provided configuration.
func NewGenericReconciler[K comparable, T any](cfg GenericReconcilerConfig[K, T]) (*GenericReconciler[K, T], error) {
	if err := cfg.CheckAndSetDefaults(); err != nil {
		return nil, trace.Wrap(err)
	}

	var val T
	kind, err := types.GetKind(val)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return &GenericReconciler[K, T]{
		cfg:    cfg,
		logger: cfg.Logger.With("kind", kind),
	}, nil
}
```

Reply:
It's weird but when you do:

```
var val T
fmt.Printf("v=%#v v.(type)=%T\n", val)
```

it will print `v=<nil> v.(type)<nil>`.

And it doesn't happen in tests. Only when the binary is built. I don't want to spend too long to investigate that, so I'll revert to my previous changes.

Reply:
Ah the `types.GetKind` actually tries to get the filed value where initially I thought that this is  just a type kind. 
Sorry for confusion.